{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Per-gene lists of participants with splice variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from cvfgaou import hailtools, gctools, notation, data\n",
    "from cvfgaou.notation import GEQ_CHAR, LEQ_CHAR\n",
    "\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]\n",
    "\n",
    "RESULT_DIR = f'{BUCKET}/spliceAI_2025-12-10'\n",
    "SPLICE_AI_FILTER_THRESHOLD = 0.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "hl.init()\n",
    "wgs_mt_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\")\n",
    "wgs_mt_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load WGS\n",
    "wgs_mt = hl.read_matrix_table(wgs_mt_path)\n",
    "wgs_mt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the VAT to as SpliceAI source\n",
    "\n",
    "for gene in tqdm(data.gene_phenotypes):\n",
    "    \n",
    "    splice_carriers_file = f'{RESULT_DIR}/splice_carriers_{gene}.parquet'\n",
    "    \n",
    "    if gctools.blob_exists(splice_carriers_file): continue\n",
    "    \n",
    "    vat_col_types={\n",
    "        'contig': str,\n",
    "        'position': int,\n",
    "        'ref_allele': str,\n",
    "        'alt_allele': str,\n",
    "        'splice_ai_acceptor_gain_score': float,\n",
    "        'splice_ai_acceptor_loss_score': float,\n",
    "        'splice_ai_donor_gain_score': float,\n",
    "        'splice_ai_donor_loss_score': float\n",
    "    }\n",
    "    \n",
    "    gene_vat = pd.read_table(\n",
    "        f'{BUCKET}/aux_data/{gene}_vat.tsv',\n",
    "        usecols=vat_col_types.keys(),\n",
    "        dtype=vat_col_types\n",
    "    )\n",
    "    \n",
    "    # Filter SpliceAI score\n",
    "    gene_vat = gene_vat[\n",
    "        (gene_vat.splice_ai_acceptor_gain_score > SPLICE_AI_FILTER_THRESHOLD) |\n",
    "        (gene_vat.splice_ai_acceptor_loss_score > SPLICE_AI_FILTER_THRESHOLD) |\n",
    "        (gene_vat.splice_ai_donor_gain_score > SPLICE_AI_FILTER_THRESHOLD) |\n",
    "        (gene_vat.splice_ai_donor_loss_score > SPLICE_AI_FILTER_THRESHOLD)\n",
    "    ]\n",
    "    \n",
    "    if not gene_vat.empty:\n",
    "    \n",
    "        try:\n",
    "            persons_df = hailtools.get_cols_with_variants(\n",
    "                gene_vat,\n",
    "                wgs_mt,\n",
    "                contig_col='contig',\n",
    "                pos_col='position',\n",
    "                ref_col='ref_allele',\n",
    "                alt_col='alt_allele'\n",
    "            )\n",
    "        except:\n",
    "            print('Failed on')\n",
    "            print(gene_vat)\n",
    "            raise\n",
    "    \n",
    "    persons_df.to_parquet(splice_carriers_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
