{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "from cvfgaou import hailtools, gctools\n",
    "\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes of interest\n",
    "genes = [\n",
    "    'APP',\n",
    "    'BAP1',\n",
    "    'BARD1',\n",
    "    'BRCA1',\n",
    "    'BRCA2',\n",
    "    'BRIP1',\n",
    "    'CALM1',\n",
    "    'CALM2',\n",
    "    'CALM3',\n",
    "    'GCK',\n",
    "    'KCNH2',\n",
    "    'KCNQ4',\n",
    "    'MSH2',\n",
    "    'OTC',\n",
    "    'PALB2',\n",
    "    'PRKN',\n",
    "    'PTEN',\n",
    "    'RAD51C',\n",
    "    'RAD51D',\n",
    "    'SCN5A',\n",
    "    'SNCA',\n",
    "    'TARDBP',\n",
    "    'TP53',\n",
    "    'VWF'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "hl.init()\n",
    "wgs_mt_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\")\n",
    "wgs_mt_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load wgs\n",
    "wgs_mt = hl.read_matrix_table(wgs_mt_path)\n",
    "wgs_mt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary for filtering VAT variants.\n",
    "# form is label: (inclusion substring, [list of exclusion substrings])\n",
    "\n",
    "vat_filters = {\n",
    "    'Pathogenic': ('pathogenic', ['likely', 'benign', 'uncertain']),\n",
    "    'Likely Pathogenic': ('pathogenic', ['benign', 'uncertain']),\n",
    "    'Uncertain Significance': ('uncertain', []),\n",
    "    'Likely Benign': ('benign', ['pathogenic', 'uncertain']),\n",
    "    'Benign': ('benign', ['likely', 'pathogenic', 'uncertain'])\n",
    "}\n",
    "\n",
    "for gene in tqdm(genes):\n",
    "    \n",
    "    out_file = f'{BUCKET}/prelim_exposures/clinvar_vat_{gene}.csv'\n",
    "    \n",
    "    if gctools.blob_exists(out_file): continue\n",
    "    \n",
    "    gene_result_dfs = []\n",
    "    \n",
    "    gene_vat = pd.read_table(f'{BUCKET}/aux_data/{gene.lower()}_vat.tsv')\n",
    "    gene_vat = gene_vat[~gene_vat.clinvar_classification.isna()]\n",
    "    \n",
    "    for classification, (inclusion_str, exclusion_strs) in tqdm(vat_filters.items()):\n",
    "        \n",
    "        variant_selection = gene_vat.clinvar_classification.str.contains(inclusion_str)\n",
    "        for ex_str in exclusion_strs:\n",
    "            variant_selection &= ~gene_vat.clinvar_classification.str.contains(ex_str)\n",
    "        \n",
    "        variant_df = gene_vat[['contig', 'position', 'ref_allele', 'alt_allele']][\n",
    "            variant_selection\n",
    "        ]\n",
    "        \n",
    "        if variant_df.empty: continue\n",
    "        \n",
    "        result_df = hailtools.get_cols_with_variants(\n",
    "            variant_df,\n",
    "            wgs_mt,\n",
    "            contig_col='contig',\n",
    "            pos_col='position',\n",
    "            ref_col='ref_allele',\n",
    "            alt_col='alt_allele'\n",
    "        )\n",
    "        \n",
    "        result_df.rename(columns = {'s': 'person_id'}, inplace=True)\n",
    "        result_df['Dataset'] = 'ClinVar'\n",
    "        result_df['Gene'] = gene\n",
    "        result_df['Classifier'] = 'Curated (AoU CDR v8 VAT)'\n",
    "        result_df['Classification'] = classification\n",
    "        \n",
    "        gene_result_dfs.append(result_df)\n",
    "    \n",
    "    if gene_result_dfs:\n",
    "        pd.concat(gene_result_dfs, ignore_index=True).to_csv(out_file, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
