{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from cvfgaou import hailtools, gctools\n",
    "\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]\n",
    "SPLICE_AI_FILTER_MAX = 0.2\n",
    "#AF_FILTER_MAX = 0.01 # We don't filter for rare variants in functional data\n",
    "DATAFRAME_VERSION = '17796333'\n",
    "RESULTS_DIR = f'{BUCKET}/classes_2025-12-15'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "hl.init()\n",
    "wgs_mt_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\")\n",
    "wgs_mt_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load wgs\n",
    "wgs_mt = hl.read_matrix_table(wgs_mt_path)\n",
    "wgs_mt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil ls $WORKSPACE_BUCKET/cvfg_17796333"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CVFG dataframe\n",
    "cvfg_df = pd.read_csv(\n",
    "    f'{BUCKET}/cvfg_17796333/final_pillar_data_with_clinvar_18_25_gnomad_wREVEL_wAM_wspliceAI_wMutpred2_wtrainvar_gold_standards_expanded_111225.csv.gz',\n",
    "    dtype={\n",
    "        'Dataset': str,\n",
    "        'Gene': str,\n",
    "        'Chrom': str,\n",
    "        #'hg38_start': int,\n",
    "        #'hg38_end': int,\n",
    "        'ref_allele': str,\n",
    "        'alt_allele': str,\n",
    "        #'auth_reported_score': float,\n",
    "        'auth_reported_func_class': str\n",
    "    }\n",
    ")\n",
    "cvfg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check author provided vales on our genes\n",
    "working_df = cvfg_df[\n",
    "    cvfg_df.Gene.isin(data.gene_phenotypes) &\n",
    "    (cvfg_df.Flag != '*')# Drop flagged variants (mapping errors etc.)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_bins_df = pd.read_csv(f'{BUCKET}/clinvar/clinvar-bins.csv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene, per_gene_df in tqdm(working_df.groupby('Gene')):\n",
    "    \n",
    "    exposures_file = f'{RESULTS_DIR}/exposures/author-reported_{gene}.parquet'\n",
    "    clinvar_file = f'{RESULTS_DIR}/clinvar_maps/author-reported_{gene}.parquet'\n",
    "    af_file = f'{RESULTS_DIR}/af_maps/author-reported_{gene}.parquet'\n",
    "    \n",
    "    if gctools.blob_exists(exposures_file): continue\n",
    "     \n",
    "    gene_result_dfs = []\n",
    "    clinvar_classes_dfs = []\n",
    "    joint_af_map = {}\n",
    "    \n",
    "    for (dataset, functional_class), df in tqdm(per_gene_df.groupby(['Dataset', 'auth_reported_func_class'])):\n",
    "\n",
    "        try:\n",
    "            variant_df = df[\n",
    "                ['Chrom', 'hg38_start', 'ref_allele', 'alt_allele']\n",
    "            ].dropna(\n",
    "                how='any',\n",
    "            ).assign(\n",
    "                Chromosome = lambda df: 'chr'+df['Chrom']\n",
    "            ).astype({'hg38_start': int})\n",
    "        except:\n",
    "            print(f'Failed while making df for {dataset}, {gene}, {functional_class}')\n",
    "            raise\n",
    "        \n",
    "        if variant_df.empty: continue\n",
    "            \n",
    "        try:\n",
    "            exposure_df, af_map, clinvar_df = hailtools.get_exposure_package(\n",
    "                variant_df,\n",
    "                wgs_mt,\n",
    "                clinvar_bins_df,\n",
    "                contig_col='Chromosome',\n",
    "                pos_col='hg38_start',\n",
    "                ref_col='ref_allele',\n",
    "                alt_col='alt_allele',\n",
    "                metadata_dict={\n",
    "                    'Dataset': dataset,\n",
    "                    'Gene': gene,\n",
    "                    'Classifier': 'Author Reported',\n",
    "                    'Classification': functional_class,\n",
    "                    'Data Version': DATAFRAME_VERSION\n",
    "                    'Measures splicing': (df.splice_measure == 'Yes').all()\n",
    "                })\n",
    "            )\n",
    "            \n",
    "            clinvar_classes_dfs.append(clinvar_df)\n",
    "            joint_af_map.update(af_map)\n",
    "            gene_result_dfs.append(exposure_df)\n",
    "        except:\n",
    "            print(f'Failed on {dataset}, {gene}, {functional_class}:')\n",
    "            print(variant_df)\n",
    "            raise\n",
    "        \n",
    "    if clinvar_classes_dfs:\n",
    "        pd.concat(clinvar_classes_dfs, ignore_index=True).to_parquet(clinvar_file, index=False)\n",
    "    if joint_af_map:\n",
    "        pd.Series(joint_af_map).to_frame(name='AF').to_parquet(af_file)\n",
    "    if gene_result_dfs:\n",
    "        pd.concat(gene_result_dfs, ignore_index=True).to_parquet(exposures_file, index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
