{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Per-gene exposure tables for MutPred2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from cvfgaou import hailtools, gctools, notation, data\n",
    "from cvfgaou.notation import GEQ_CHAR, LEQ_CHAR\n",
    "\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]\n",
    "\n",
    "RESULT_DIR = f'{BUCKET}/classes_2025-10-15'\n",
    "SPLICE_AI_FILTER_THRESHOLD = np.inf#0.2\n",
    "AF_FILTER_THRESHOLD = np.inf#0.01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "hl.init()\n",
    "wgs_mt_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\")\n",
    "wgs_mt_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load WGS\n",
    "wgs_mt = hl.read_matrix_table(wgs_mt_path)\n",
    "wgs_mt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load mutpred table and filter it down to the genes of interest\n",
    "mutpred_df = pd.concat(\n",
    "    [\n",
    "        df[df['gene_symbol'].isin(data.project_genes)]\n",
    "        for df in tqdm(pd.read_table(f'{BUCKET}/mutpred2/IGVFFI7749UFOK.tsv.gz', chunksize=100000))\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract AA change columns\n",
    "mutpred_df[['aa_ref', 'aa_pos', 'aa_alt']] = mutpred_df['Substitution'].str.extract(\n",
    "    r'^([A-Z])(\\d+)([A-Z])$'\n",
    ")\n",
    "mutpred_df = mutpred_df.assign(\n",
    "    aa_ref=lambda df: df.aa_ref.map(notation.AA_NAMES),\n",
    "    aa_alt=lambda df: df.aa_alt.map(notation.AA_NAMES)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene-specific Calibration table\n",
    "# This table is indexed by gene\n",
    "mp2_gene_thresholds_df = pd.read_csv(\n",
    "    f'{BUCKET}/calibrations/pillarg_MP2_gene_specific_thresh.csv',\n",
    "    index_col=0\n",
    ")\n",
    "\n",
    "# Mapping from our thresholds to thresholds used in the table and the respective comparison direction\n",
    "gs_threshold_map = {\n",
    "    f'{inequality} {sign}{points}': (f'{criterion}_{strength.title()}', comparator)\n",
    "    for inequality, sign, criterion, _, comparator, _ in notation.DOE_TABLE\n",
    "    for points, strength in notation.SOE_TABLE\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_bins_df = pd.read_csv(f'{BUCKET}/clinvar/clinvar-bins.csv.gz')\n",
    "clinvar_bins_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the VAT to match AA changes to variants\n",
    "\n",
    "for gene, per_gene_df in tqdm(mutpred_df.groupby('gene_symbol')):\n",
    "    \n",
    "    exposures_file = f'{RESULT_DIR}/exposures/mutpred2_{gene}.parquet'\n",
    "    clinvar_file = f'{RESULT_DIR}/clinvar_maps/mutpred2_{gene}.parquet'\n",
    "    af_file = f'{RESULT_DIR}/af_maps/mutpred2_{gene}.parquet'\n",
    "    \n",
    "    if gctools.blob_exists(exposures_file): continue\n",
    "     \n",
    "    gene_result_dfs = []\n",
    "    clinvar_classes_dfs = []\n",
    "    joint_af_map = {}\n",
    "    \n",
    "    # Calibration thresholds from Bergquist et al. 10.1016/j.gim.2025.101402\n",
    "    \n",
    "    variant_classes = (\n",
    "        (c, per_gene_df[selection])\n",
    "        for c, selection in (\n",
    "            (f'{LEQ_CHAR} -4', per_gene_df['MutPred2 score'] <= 0.010), # Benign Strong\n",
    "            (f'{LEQ_CHAR} -3', per_gene_df['MutPred2 score'] <= 0.031), # Benign Moderate+\n",
    "            (f'{LEQ_CHAR} -2', per_gene_df['MutPred2 score'] <= 0.197), # Benign Moderate\n",
    "            (f'{LEQ_CHAR} -1', per_gene_df['MutPred2 score'] <= 0.391), # Benign Supporting\n",
    "            (f'{GEQ_CHAR} +1', per_gene_df['MutPred2 score'] >= 0.737), # Pathogenic Supporting\n",
    "            (f'{GEQ_CHAR} +2', per_gene_df['MutPred2 score'] >= 0.829), # Pathogenic Moderate\n",
    "            (f'{GEQ_CHAR} +3', per_gene_df['MutPred2 score'] >= 0.895), # Pathogenic Moderate+\n",
    "            (f'{GEQ_CHAR} +4', per_gene_df['MutPred2 score'] >= 0.932)  # Pathogenic Strong\n",
    "        )\n",
    "    )\n",
    "\n",
    "    gene_vat = pd.read_table(f'{BUCKET}/aux_data/{gene}_vat.tsv')\n",
    "    \n",
    "    # Filter SpliceAI score and AF\n",
    "    gene_vat = gene_vat[\n",
    "        (gene_vat.gnomad_all_af <= AF_FILTER_THRESHOLD) &\n",
    "        ~(\n",
    "            (gene_vat.splice_ai_acceptor_gain_score > SPLICE_AI_FILTER_THRESHOLD) |\n",
    "            (gene_vat.splice_ai_acceptor_loss_score > SPLICE_AI_FILTER_THRESHOLD) |\n",
    "            (gene_vat.splice_ai_donor_gain_score > SPLICE_AI_FILTER_THRESHOLD) |\n",
    "            (gene_vat.splice_ai_donor_loss_score > SPLICE_AI_FILTER_THRESHOLD)\n",
    "        )\n",
    "    ]\n",
    "    \n",
    "    # Match on AA change    \n",
    "    gene_vat[['aa_ref', 'aa_pos', 'aa_alt']] = gene_vat.aa_change.str.extract(\n",
    "        r'p\\.\\(([A-Z][a-z][a-z])(\\d+)([A-Z][a-z][a-z])\\)'\n",
    "    )\n",
    "    \n",
    "    for classification, mutpred_class_df in tqdm(variant_classes):\n",
    "        \n",
    "        variant_df = gene_vat.merge(\n",
    "            mutpred_class_df,\n",
    "            left_on=['aa_ref', 'aa_pos', 'aa_alt'],\n",
    "            right_on=['aa_ref', 'aa_pos', 'aa_alt'],\n",
    "            how='inner'\n",
    "        )[['contig', 'position', 'ref_allele', 'alt_allele']]\n",
    "        \n",
    "        if variant_df.empty: continue\n",
    "        \n",
    "        exposure_df, af_map, clinvar_df = hailtools.get_exposure_package(\n",
    "            variant_df,\n",
    "            wgs_mt,\n",
    "            clinvar_bins_df,\n",
    "            contig_col='contig',\n",
    "            pos_col='position',\n",
    "            ref_col='ref_allele',\n",
    "            alt_col='alt_allele',\n",
    "            metadata_dict={\n",
    "                'Dataset': 'MutPred2',\n",
    "                'Gene': gene,\n",
    "                'Classifier': 'Calibrated (Bergquist et al. 10.1016/j.gim.2025.101402)',\n",
    "                'Classification': classification,\n",
    "                'SpliceAI filter max': SPLICE_AI_FILTER_THRESHOLD,\n",
    "                'AF filter max': AF_FILTER_THRESHOLD\n",
    "            }\n",
    "        )\n",
    "        \n",
    "        clinvar_classes_dfs.append(clinvar_df)\n",
    "        joint_af_map.update(af_map)\n",
    "        gene_result_dfs.append(exposure_df)\n",
    "    \n",
    "    # Gene-specific AM2 thresholds\n",
    "    if gene in mp2_gene_thresholds_df.index:\n",
    "        for classification, (label, compare) in gs_threshold_map.items():\n",
    "            threshold = mp2_gene_thresholds_df.loc[gene, label]\n",
    "            if np.isnan(threshold):\n",
    "                continue\n",
    "            \n",
    "            mutpred_class_df = per_gene_df[compare(per_gene_df['MutPred2 score'], threshold)]\n",
    "            \n",
    "            variant_df = gene_vat.merge(\n",
    "            mutpred_class_df,\n",
    "                left_on=['aa_ref', 'aa_pos', 'aa_alt'],\n",
    "                right_on=['aa_ref', 'aa_pos', 'aa_alt'],\n",
    "                how='inner'\n",
    "            )[['contig', 'position', 'ref_allele', 'alt_allele']]\n",
    "\n",
    "            if variant_df.empty: continue\n",
    "\n",
    "            exposure_df, af_map, clinvar_df = hailtools.get_exposure_package(\n",
    "                variant_df,\n",
    "                wgs_mt,\n",
    "                clinvar_bins_df,\n",
    "                contig_col='contig',\n",
    "                pos_col='position',\n",
    "                ref_col='ref_allele',\n",
    "                alt_col='alt_allele',\n",
    "                metadata_dict = {\n",
    "                    'Dataset': 'MutPred2',\n",
    "                    'Gene': gene,\n",
    "                    'Classifier': 'Calibrated (gene-specific)',\n",
    "                    'Classification': classification,\n",
    "                    'SpliceAI filter max': SPLICE_AI_FILTER_THRESHOLD,\n",
    "                    'AF filter max': AF_FILTER_THRESHOLD\n",
    "                }\n",
    "            )\n",
    "\n",
    "            clinvar_classes_dfs.append(clinvar_df)\n",
    "            joint_af_map.update(af_map)\n",
    "            gene_result_dfs.append(exposure_df)\n",
    "    \n",
    "    if clinvar_classes_dfs:\n",
    "        pd.concat(clinvar_classes_dfs, ignore_index=True).to_parquet(clinvar_file, index=False)\n",
    "    if joint_af_map:\n",
    "        pd.Series(joint_af_map).to_frame(name='AF').to_parquet(af_file)\n",
    "    if gene_result_dfs:\n",
    "        pd.concat(gene_result_dfs, ignore_index=True).to_parquet(exposures_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
