{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from cvfgaou import hailtools, gctools\n",
    "\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes of interest\n",
    "genes = [\n",
    "    'APP',\n",
    "    'BAP1',\n",
    "    'BARD1',\n",
    "    'BRCA1',\n",
    "    'BRCA2',\n",
    "    'BRIP1',\n",
    "    'CALM1',\n",
    "    'CALM2',\n",
    "    'CALM3',\n",
    "    'GCK',\n",
    "    'KCNH2',\n",
    "    'KCNQ4',\n",
    "    'MSH2',\n",
    "    'OTC',\n",
    "    'PALB2',\n",
    "    'PRKN',\n",
    "    'PTEN',\n",
    "    'RAD51C',\n",
    "    'RAD51D',\n",
    "    'SCN5A',\n",
    "    'SNCA',\n",
    "    'TARDBP',\n",
    "    'TP53',\n",
    "    'VWF'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "hl.init()\n",
    "wgs_mt_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\")\n",
    "wgs_mt_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load wgs\n",
    "wgs_mt = hl.read_matrix_table(wgs_mt_path)\n",
    "wgs_mt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bergquist et al. thresholds 10.1016/j.gim.2025.101402\n",
    "\n",
    "# Benign Moderate: <= 0.199\n",
    "# Benign Supporting: <= 0.170\n",
    "# Pathogenic Supporting: >= 0.792\n",
    "# Pathogenic Moderate: >= 0.906\n",
    "# Pathogenic Strong: >= 0.990"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since the AM file is large, we will take a different approach here:\n",
    "# we will iterate through the file and populate our gene dataframes.\n",
    "\n",
    "am_df = pd.read_table(\n",
    "    f'{BUCKET}/precomputed/AlphaMissense_hg38.tsv.gz',\n",
    "    header=3,\n",
    "    #engine='pyarrow',\n",
    "    chunksize=500000\n",
    ")\n",
    "\n",
    "# AM columns are:\n",
    "# CHROM, POS, REF, ALT, genome, uniprot_id, transcript_id, protein_variant, am_pathogenicity, am_class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need the interval spec of each gene:\n",
    "# Obtain genomic coordinates\n",
    "\n",
    "intervals_df = pd.read_table(\n",
    "    f'{BUCKET}/aux_data/gene_metadata.txt',\n",
    "    usecols=['Gene stable ID version', 'Chromosome/scaffold name', 'Gene start (bp)', 'Gene end (bp)', 'Gene name'],\n",
    "    dtype=str\n",
    ").drop_duplicates()\n",
    "\n",
    "# Filter to genes of interest\n",
    "\n",
    "intervals_df = intervals_df[intervals_df['Gene name'].isin(genes)]\n",
    "\n",
    "# Cleanup\n",
    "\n",
    "intervals_df = intervals_df[intervals_df['Chromosome/scaffold name'] != 'HG2334_PATCH']\n",
    "\n",
    "# Nicer notation\n",
    "intervals_df['Chrom'] = 'chr' + intervals_df['Chromosome/scaffold name']\n",
    "intervals_df = intervals_df.astype(\n",
    "    {\n",
    "        'Gene start (bp)': int,\n",
    "        'Gene end (bp)': int\n",
    "    }\n",
    ")\n",
    "\n",
    "# Check\n",
    "intervals_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "am_gene_df_lists = defaultdict(list)\n",
    "\n",
    "for chunk in tqdm(am_df):\n",
    "    select_intervals_df = intervals_df[\n",
    "        intervals_df['Chrom'].isin(chunk['#CHROM'].drop_duplicates())\n",
    "    ][['Gene start (bp)', 'Gene end (bp)', 'Gene name', 'Chrom']]\n",
    "    \n",
    "    for start, end, gene, chrom in select_intervals_df.itertuples(index=False):\n",
    "        am_gene_df_lists[gene].append(chunk[\n",
    "            (chunk['#CHROM'] == chrom) &\n",
    "            (chunk['POS'] >= start) &\n",
    "            (chunk['POS'] <= end)\n",
    "        ])\n",
    "\n",
    "am_gene_dfs = {\n",
    "    gene: pd.concat(df_list, ignore_index=True)\n",
    "    for gene, df_list in tqdm(am_gene_df_lists.items())\n",
    "    if df_list\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene-specific Calibration table\n",
    "# This table is indexed by gene\n",
    "am_gene_thresholds_df = pd.read_csv(\n",
    "    f'{BUCKET}/calibrations/pillarg_AM_gene_specific_thresh.csv',\n",
    "    index_col=0\n",
    ")\n",
    "\n",
    "# Mapping from our thresholds to thresholds used in the table and the respective comparison direction\n",
    "gs_threshold_map = {\n",
    "    f'{direction} {strength}': (f'{direction_label}_{strength.title()}', comparator)\n",
    "    for direction, direction_label, comparator in (\n",
    "        ('Pathogenic', 'PP3', pd.Series.ge),\n",
    "        ('Benign', 'BP4', pd.Series.le)\n",
    "    )\n",
    "    for strength in ('very strong', 'strong', 'moderate', 'supporting')\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_bins_df = pd.read_csv(f'{BUCKET}/clinvar/clinvar-bins.csv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import chain\n",
    "\n",
    "for gene, per_gene_df in tqdm(am_gene_dfs.items()):\n",
    "    \n",
    "    exposures_file = f'{BUCKET}/classes_2025-07-22/exposures/alphamissense_{gene}.parquet'\n",
    "    clinvar_file = f'{BUCKET}/classes_2025-07-22/clinvar_maps/alphamissense_{gene}.parquet'\n",
    "    af_file = f'{BUCKET}/classes_2025-07-22/af_maps/alphamissense_{gene}.parquet'\n",
    "    \n",
    "    if gctools.blob_exists(exposures_file): continue\n",
    "     \n",
    "    gene_result_dfs = []\n",
    "    clinvar_classes_dfs = []\n",
    "    joint_af_map = {}\n",
    "    \n",
    "    variant_classes = chain(\n",
    "        ( # Author Reported scores; also, make the label text nicer\n",
    "            ('Author Reported', c.replace('_', ' ').capitalize(), df)\n",
    "            for c, df in per_gene_df.groupby('am_class')\n",
    "        ),\n",
    "        ( # Calibrated scores\n",
    "            ('Calibrated (Bergquist et al. 10.1016/j.gim.2025.101402)', c, per_gene_df[selection])\n",
    "            for c, selection in (\n",
    "                ('Benign moderate', per_gene_df['am_pathogenicity'] <= 0.199),\n",
    "                ('Benign supporting', per_gene_df['am_pathogenicity'] <= 0.170),\n",
    "                ('Pathogenic supporting', per_gene_df['am_pathogenicity'] >= 0.792),\n",
    "                ('Pathogenic moderate', per_gene_df['am_pathogenicity'] >= 0.906),\n",
    "                ('Pathogenic strong', per_gene_df['am_pathogenicity'] >= 0.990)\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    # Gene-specific hresholds\n",
    "    if gene in am_gene_thresholds_df.index:\n",
    "        variant_classes = chain(\n",
    "            variant_classes,\n",
    "            (\n",
    "                ('Calibrated (gene-specific)', c, per_gene_df[selection])\n",
    "                for c, selection in (\n",
    "                    (classification, compare(per_gene_df['am_pathogenicity'], threshold))\n",
    "                    for classification, (label, compare) in gs_threshold_map.items()\n",
    "                    for threshold in (am_gene_thresholds_df.loc[gene, label], )\n",
    "                    if not np.isnan(threshold)\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    for classifier, classification, variant_df in tqdm(variant_classes):\n",
    "\n",
    "        if variant_df.empty: continue\n",
    "            \n",
    "        try:\n",
    "            exposure_df, af_map, clinvar_df = hailtools.get_exposure_package(\n",
    "                variant_df,\n",
    "                wgs_mt,\n",
    "                clinvar_bins_df,\n",
    "                contig_col='#CHROM',\n",
    "                pos_col='POS',\n",
    "                ref_col='REF',\n",
    "                alt_col='ALT',\n",
    "                metadata_dict = {\n",
    "                    'Dataset': 'AlphaMissense',\n",
    "                    'Gene': gene,\n",
    "                    'Classifier': classifier,\n",
    "                    'Classification': classification\n",
    "                }\n",
    "            )\n",
    "\n",
    "            clinvar_classes_dfs.append(clinvar_df)\n",
    "            joint_af_map.update(af_map)\n",
    "            gene_result_dfs.append(exposure_df)\n",
    "            \n",
    "        except:\n",
    "            print(f'Failed on {gene}, {classifier}, {classification}:')\n",
    "            print(variant_df)\n",
    "            raise\n",
    "        \n",
    "    if clinvar_classes_dfs:\n",
    "        pd.concat(clinvar_classes_dfs, ignore_index=True).to_parquet(clinvar_file, index=False)\n",
    "    if joint_af_map:\n",
    "        pd.Series(joint_af_map).to_frame(name='AF').to_parquet(af_file)\n",
    "    if gene_result_dfs:\n",
    "        pd.concat(gene_result_dfs, ignore_index=True).to_parquet(exposures_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
