{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "import json\n",
    "\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from cvfgaou import gctools\n",
    "\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes of interest\n",
    "genes = [\n",
    "    'APP',\n",
    "    'BAP1',\n",
    "    'BARD1',\n",
    "    'BRCA1',\n",
    "    'BRCA2',\n",
    "    'BRIP1',\n",
    "    'CALM1',\n",
    "    'CALM2',\n",
    "    'CALM3',\n",
    "    'GCK',\n",
    "    'KCNH2',\n",
    "    'KCNQ4',\n",
    "    'MSH2',\n",
    "    'OTC',\n",
    "    'PALB2',\n",
    "    'PRKN',\n",
    "    'PTEN',\n",
    "    'RAD51C',\n",
    "    'RAD51D',\n",
    "    'SCN5A',\n",
    "    'SNCA',\n",
    "    'TARDBP',\n",
    "    'TP53',\n",
    "    'VWF'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_types = {\n",
    "    'Type': str,\n",
    "    'GeneSymbol': str,\n",
    "    'ClinicalSignificance': str,\n",
    "    'Assembly': str,\n",
    "    'Chromosome': str,\n",
    "    'PositionVCF': int,\n",
    "    'ReferenceAlleleVCF': str,\n",
    "    'AlternateAlleleVCF': str\n",
    "}\n",
    "\n",
    "clinvar_df = pd.read_table(\n",
    "    f'{BUCKET}/clinvar/variant_summary_2025-01.txt.gz',\n",
    "    usecols = list(col_types.keys()),\n",
    "    dtype = col_types,\n",
    "    engine = 'pyarrow'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter to:\n",
    "# -GRCh38\n",
    "# -genes of interest\n",
    "# -non-null Alleles\n",
    "clinvar_df.dropna(subset=['PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF'], how='any', inplace=True)\n",
    "clinvar_df = clinvar_df[\n",
    "    (clinvar_df.PositionVCF > 0) &\n",
    "    (clinvar_df.Assembly == 'GRCh38') &\n",
    "    (clinvar_df.GeneSymbol.isin(genes))\n",
    "]\n",
    "# Make chromosome notation consistent with Hail\n",
    "clinvar_df['Chromosome'] = 'chr' + clinvar_df['Chromosome']\n",
    "clinvar_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_df.ClinicalSignificance.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build table of ClinVar bins that will be used to annotate other classifications"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will bin into the following mutex significance categories:\n",
    "\n",
    "- Benign\n",
    "- Likely benign, inclues\n",
    "    - Likely benign\n",
    "    - Benign/Likely benign\n",
    "- Uncertain significance, includes\n",
    "    - Uncertain significance\n",
    "    - Uncertain significance/Uncertain risk allele\n",
    "- Conflicting\n",
    "    - Conflicting classifications of pathogenicity\n",
    "    - Conflicting classifications of pathogenicity; risk factor\n",
    "- Likely pathogenic, includes\n",
    "    - Likely pathogenic\n",
    "    - Pathogenic/Likely pathogenic\n",
    "    - Likely pathogenic/Likely risk allele\n",
    "    - Pathogenic/Likely pathogenic/Likely risk allele\n",
    "- Pathogenic\n",
    "- Other / not in ClinVar\n",
    "    - All other classifications.\n",
    "\n",
    "Notably, \"Other\" will also include variants not in ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutex_binning = {\n",
    "    'Pathogenic': ['Pathogenic'],\n",
    "    'Likely pathogenic': [\n",
    "        'Likely pathogenic',\n",
    "        'Pathogenic/Likely pathogenic',\n",
    "        'Likely pathogenic/Likely risk allele',\n",
    "        'Pathogenic/Likely pathogenic/Likely risk allele',\n",
    "    ],\n",
    "    'Uncertain significance': [\n",
    "        'Uncertain significance',\n",
    "        'Uncertain significance/Uncertain risk allele'\n",
    "    ],\n",
    "    'Conflicting': [\n",
    "        'Conflicting classifications of pathogenicity',\n",
    "        'Conflicting classifications of pathogenicity; risk factor'\n",
    "    ],\n",
    "    'Likely benign': [\n",
    "        'Likely benign',\n",
    "        'Benign/Likely benign'\n",
    "    ],\n",
    "    'Benign': ['Benign']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build a table that defines variant: ClinVar bin\n",
    "sig2bin_map = {\n",
    "    significance: sig_bin\n",
    "    for sig_bin, values in mutex_binning.items()\n",
    "    for significance in values\n",
    "}\n",
    "\n",
    "clinvar_bins_df = clinvar_df[['Chromosome', 'PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF']].copy()\n",
    "clinvar_bins_df['Clinvar significance'] = clinvar_df.ClinicalSignificance.map(sig2bin_map)\n",
    "clinvar_bins_df.dropna(axis='index', how='any', inplace=True)\n",
    "clinvar_bins_df.rename(\n",
    "    columns={\n",
    "        'PositionVCF': 'Pos',\n",
    "        'ReferenceAlleleVCF': 'Ref',\n",
    "        'AlternateAlleleVCF': 'Alt'\n",
    "    },\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_variants_with_clinvar(\n",
    "    variant_df,\n",
    "    chrom_col,\n",
    "    pos_col,\n",
    "    ref_col,\n",
    "    alt_col,\n",
    "    clinvar_bins_df=clinvar_bins_df\n",
    "):\n",
    "    varspec = ['Chromosome', 'Pos', 'Ref', 'Alt']\n",
    "    df = variant_df[[chrom_col, pos_col, ref_col, alt_col]]\n",
    "    df.columns = varspec\n",
    "    result_df = df.merge(\n",
    "        clinvar_bins_df,\n",
    "        on=varspec,\n",
    "        how='left'\n",
    "    )\n",
    "    result_df['Clinvar significance'] = result_df['Clinvar significance'].fillna('Other / not in ClinVar')\n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get variant sets for all classifications"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ClinVar class groups\n",
    "\n",
    "For OR classes, we will use the following significance categories:\n",
    "\n",
    "- Benign\n",
    "- Likely benign, inclues\n",
    "    - Benign\n",
    "    - Likely benign\n",
    "    - Benign/Likely benign\n",
    "- Uncertain significance, includes\n",
    "    - Uncertain significance\n",
    "    - Uncertain significance/Uncertain risk allele\n",
    "- Likely pathogenic, includes\n",
    "    - Pathogenic\n",
    "    - Likely pathogenic\n",
    "    - Pathogenic/Likely pathogenic\n",
    "    - Likely pathogenic/Likely risk allele\n",
    "    - Pathogenic/Likely pathogenic/Likely risk allele\n",
    "- Pathogenic\n",
    "\n",
    "Notably, categories are not mutually exclusive, and we drop \"uncertain risk allele,\" \"conflicting,\" and \"not provided\" variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "binning = {\n",
    "    'Pathogenic': ['Pathogenic'],\n",
    "    'Likely pathogenic': [\n",
    "        'Pathogenic',\n",
    "        'Likely pathogenic',\n",
    "        'Pathogenic/Likely pathogenic',\n",
    "        'Likely pathogenic/Likely risk allele',\n",
    "        'Pathogenic/Likely pathogenic/Likely risk allele',\n",
    "    ],\n",
    "    'Uncertain significance': [\n",
    "        'Uncertain significance',\n",
    "        'Uncertain significance/Uncertain risk allele'\n",
    "    ],\n",
    "    'Likely benign': [\n",
    "        'Benign',\n",
    "        'Likely benign',\n",
    "        'Benign/Likely benign'\n",
    "    ],\n",
    "    'Benign': ['Benign']\n",
    "}\n",
    "\n",
    "subset_clinvar_df = clinvar_df\n",
    "\n",
    "for label, classifications in binning.items():\n",
    "    selection = False\n",
    "    for c in classifications:\n",
    "        selection |= (subset_clinvar_df.ClinicalSignificance == c)\n",
    "\n",
    "    subset_clinvar_df[label] = selection\n",
    "        \n",
    "subset_clinvar_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_dfs = []\n",
    "\n",
    "for gene in tqdm(genes):\n",
    "    \n",
    "    for classification in tqdm(binning):\n",
    "        variant_df = subset_clinvar_df[\n",
    "            (subset_clinvar_df['GeneSymbol'] == gene) &\n",
    "            subset_clinvar_df[classification]\n",
    "        ]\n",
    "        \n",
    "        if variant_df.empty: continue\n",
    "        \n",
    "        result_df = annotate_variants_with_clinvar(\n",
    "            variant_df,\n",
    "            'Chromosome',\n",
    "            'PositionVCF',\n",
    "            'ReferenceAlleleVCF',\n",
    "            'AlternateAlleleVCF'\n",
    "        )\n",
    "        \n",
    "        result_df['Dataset'] = 'ClinVar'\n",
    "        result_df['Gene'] = gene\n",
    "        result_df['Classifier'] = 'Curated (2025-01)'\n",
    "        result_df['Classification'] = classification\n",
    "        \n",
    "        clinvar_dfs.append(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assay calibrations 2025-06-05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CVFG dataframe\n",
    "cvfg_df = pd.read_csv(\n",
    "    f'{BUCKET}/cvfg_15577805/pillar_data_clinvar38_053125_wREVEL.csv.gz',\n",
    "    index_col=0,\n",
    "    dtype={\n",
    "        'Dataset': str,\n",
    "        'Gene': str,\n",
    "        'Chrom': str,\n",
    "        #'hg38_start': int,\n",
    "        #'hg38_end': int,\n",
    "        'ref_allele': str,\n",
    "        'alt_allele': str,\n",
    "        #'auth_reported_score': float,\n",
    "        'auth_reported_func_class': str\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Limit to our genes\n",
    "working_df = cvfg_df[cvfg_df.Gene.isin(genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Evidence threshold mapping from points to labels\n",
    "evidence_strength_series = pd.Series({\n",
    "    +8: \"Pathogenic very strong\",\n",
    "    +4: \"Pathogenic strong\",\n",
    "    +2: \"Pathogenic moderate\",\n",
    "    +1: \"Pathogenic supporting\",\n",
    "    -1: \"Benign supporting\",\n",
    "    -2: \"Benign moderate\",\n",
    "    -4: \"Benign strong\",\n",
    "    -8: \"Benign very strong\"\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common container for frames\n",
    "assay_calibration_dfs = []\n",
    "\n",
    "for blob in tqdm(gctools.list_blobs(\n",
    "    f'{BUCKET}/calibrations/functional_assay_calibration_results_06_06_2025/summaries_2025_06_05/',\n",
    "    return_uris=False\n",
    ")):\n",
    "    if blob.path.endswith('.json'):\n",
    "        calibration = json.loads(blob.download_as_text())\n",
    "        \n",
    "        dataset = calibration['scoreset_name']\n",
    "        \n",
    "        dataset_df = working_df[working_df['Dataset'] == dataset]\n",
    "        \n",
    "        if dataset_df.empty:\n",
    "            continue\n",
    "        \n",
    "        # Check how thresholds are provided. If lists, assume they should be aligned to 1, 2, 3, 4, 8\n",
    "        pathogenic_thresholds = calibration['final_pathogenic_thresholds']\n",
    "        if isinstance(pathogenic_thresholds, list):\n",
    "            print(f'Warning: pathogenic thresholds provided as a list for {dataset}')\n",
    "            pathogenic_thresholds = dict(zip([1,2,3,4,8], pathogenic_thresholds))\n",
    "        \n",
    "        benign_thresholds = calibration['final_benign_thresholds']\n",
    "        if isinstance(benign_thresholds, list):\n",
    "            print(f'Warning: benign thresholds provided as a list for {dataset}')\n",
    "            benign_thresholds = dict(zip([-1,-2,-3,-4,-8], benign_thresholds))\n",
    "        \n",
    "        # Build series of thesholds\n",
    "        thresholds = pd.concat(\n",
    "            [\n",
    "                pd.Series(pathogenic_thresholds),\n",
    "                pd.Series(benign_thresholds)\n",
    "            ]\n",
    "        )\n",
    "        thresholds.index = pd.to_numeric(thresholds.index)\n",
    "        thresholds.sort_index(inplace=True)\n",
    "        \n",
    "        # Can't work without thresholds\n",
    "        if thresholds.isna().all():\n",
    "            continue\n",
    "        \n",
    "        # Directionality of thresholds\n",
    "        increasing = calibration.get('inverted') == 'inverted'\n",
    "        \n",
    "        # Check that threholds are increasing(decreasing):\n",
    "        if increasing:\n",
    "            assert thresholds.dropna().is_monotonic_increasing, f'Thresholds for {dataset} are expected to be increasing: {thresholds}'\n",
    "            comparisons = [pd.Series.le] * 4 + [pd.Series.ge] * 4\n",
    "        else:\n",
    "            assert thresholds.dropna().is_monotonic_decreasing, f'Thresholds for {dataset} are expected to be decreasing: {thresholds}'\n",
    "            comparisons = [pd.Series.ge] * 4 + [pd.Series.le] * 4\n",
    "        comparisons_series = pd.Series(comparisons, [-8,-4,-2,-1,1,2,4,8])\n",
    "        \n",
    "        # Align thresholds, labels, and comparisons\n",
    "        thresholds_df = pd.concat(\n",
    "            {\n",
    "                'Classification': evidence_strength_series,\n",
    "                'Threshold': thresholds,\n",
    "                'Comparison': comparisons_series\n",
    "            },\n",
    "            axis='columns'\n",
    "        ).dropna(how='any')\n",
    "        \n",
    "        result_dfs = []\n",
    "        \n",
    "        for classification, threshold, compare in tqdm(thresholds_df.itertuples(index=False)):\n",
    "            \n",
    "            for gene, gene_df in dataset_df.groupby('Gene'):\n",
    "                \n",
    "                variant_df = gene_df[\n",
    "                    ['Chrom', 'hg38_start', 'ref_allele', 'alt_allele']\n",
    "                ][\n",
    "                    compare(gene_df['auth_reported_score'].astype(float), threshold)\n",
    "                ].dropna(how='any').assign(Chromosome = lambda df: 'chr'+df['Chrom']).astype({'hg38_start': int})\n",
    "            \n",
    "                if variant_df.empty:\n",
    "                    continue\n",
    "\n",
    "                result_df = variant_df[['Chromosome', 'hg38_start', 'ref_allele', 'alt_allele']].rename(\n",
    "                    columns={\n",
    "                        'hg38_start': 'Pos',\n",
    "                        'ref_allele': 'Ref',\n",
    "                        'alt_allele': 'Alt'\n",
    "                    }\n",
    "                ).merge(\n",
    "                    clinvar_bins_df,\n",
    "                    on=['Chromosome', 'Pos', 'Ref', 'Alt'],\n",
    "                    how='left'\n",
    "                )\n",
    "                result_df['Clinvar significance'] = result_df['Clinvar significance'].fillna('Other / not in ClinVar')\n",
    "                \n",
    "                result_df['Dataset'] = dataset\n",
    "                result_df['Gene'] = gene\n",
    "                result_df['Classifier'] = 'Calibrated (2025-06-05)'\n",
    "                result_df['Classification'] = classification\n",
    "                \n",
    "                assay_calibration_dfs.append(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Author provided assay scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_author_dfs = []\n",
    "\n",
    "for gene, per_gene_df in tqdm(working_df.groupby('Gene')):\n",
    "    \n",
    "    for (dataset, functional_class), df in tqdm(per_gene_df.groupby(['Dataset', 'auth_reported_func_class'])):\n",
    "\n",
    "        try:\n",
    "            variant_df = df[\n",
    "                ['Chrom', 'hg38_start', 'ref_allele', 'alt_allele']\n",
    "            ].dropna(\n",
    "                how='any',\n",
    "            ).assign(\n",
    "                Chromosome = lambda df: 'chr'+df['Chrom']\n",
    "            ).astype({'hg38_start': int})\n",
    "        except:\n",
    "            print(f'Failed while making df for {dataset}, {gene}, {functional_class}')\n",
    "            raise\n",
    "        \n",
    "        if variant_df.empty: continue\n",
    "            \n",
    "        try:\n",
    "            result_df = annotate_variants_with_clinvar(\n",
    "                variant_df,\n",
    "                'Chromosome',\n",
    "                'hg38_start',\n",
    "                'ref_allele',\n",
    "                'alt_allele'\n",
    "            )\n",
    "        except:\n",
    "            print(f'Failed on {dataset}, {gene}, {functional_class}:')\n",
    "            print(variant_df)\n",
    "            raise\n",
    "        \n",
    "        result_df['Dataset'] = dataset\n",
    "        result_df['Gene'] = gene\n",
    "        result_df['Classifier'] = 'Author Reported'\n",
    "        result_df['Classification'] = functional_class\n",
    "        \n",
    "        assay_author_dfs.append(result_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AlphaMissense\n",
    "\n",
    "We will need to re-run the main workflow anyway, so we will add the variant statistics collection to that file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collect results and write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat(\n",
    "    assay_author_dfs + assay_calibration_dfs + clinvar_dfs,\n",
    "    ignore_index=True\n",
    ").to_csv(f'{BUCKET}/WIP/variant_clinvar_classes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
