{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Common imports and constants\n",
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm\n",
    "from cvfgaou import aou, odds, gctools, data\n",
    "\n",
    "DATASET = os.environ[\"WORKSPACE_CDR\"]\n",
    "BUCKET = os.environ[\"WORKSPACE_BUCKET\"]\n",
    "DATA_DIR = f'{BUCKET}/data_v2'\n",
    "CLASSES_VERSION = '2026-01-14' # Used for local dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Grab map of ENSG IDs for annotation\n",
    "\n",
    "ensg_df = pd.read_table(\n",
    "    f'{BUCKET}/aux_data/gene_metadata.txt',\n",
    "    usecols=['Gene stable ID version', 'Chromosome/scaffold name', 'Gene name'],\n",
    "    dtype=str\n",
    ").drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out to chromosomes only\n",
    "ensg_df = ensg_df[~ensg_df['Chromosome/scaffold name'].str.endswith('PATCH')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensg_filtered_df = ensg_df[ensg_df['Gene name'].isin(data.gene_phenotypes)]\n",
    "ensg_map = ensg_filtered_df['Gene stable ID version']\n",
    "ensg_map.index = ensg_filtered_df['Gene name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Grab table that describes which assays measure splicing\n",
    "splicing_map_df = pd.read_csv(\n",
    "    f'{BUCKET}/cvfg_17796333/final_pillar_data_with_clinvar_18_25_gnomad_wREVEL_wAM_wspliceAI_wMutpred2_wtrainvar_gold_standards_expanded_111225.csv.gz',\n",
    "    usecols = ['Dataset', 'splice_measure']\n",
    ").drop_duplicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "splicing_map = pd.Series(splicing_map_df['splice_measure'], index=splicing_map_df['Dataset'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene case pools definitions\n",
    "# Taking them from the gene phenotypes defined for the project\n",
    "\n",
    "cohorts = aou.CohortLoader(\n",
    "    gene_cohort_map = {\n",
    "        gene: (\n",
    "            {f'{p} case pool.tsv.gz' for p in case_phenos},\n",
    "            {f'{p} control pool.tsv.gz' for p in control_phenos}\n",
    "        )\n",
    "        for gene, (case_phenos, control_phenos, _) in data.gene_phenotypes.items()\n",
    "    },\n",
    "    ancestry_df = pd.read_table(f'{DATA_DIR}/ancestry_pca.tsv.gz', index_col='research_id'),\n",
    "    demo_df = pd.read_table(f'{DATA_DIR}/demo.tsv.gz', index_col='person_id'),\n",
    "    data_dir = DATA_DIR\n",
    ")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil ls $WORKSPACE_BUCKET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p classes_2026-01-14\n",
    "# Calibrated assay scores\n",
    "!gsutil -m cp -r $WORKSPACE_BUCKET/classes_2026-01-13/* classes_2026-01-14/\n",
    "# Author reported assay scores\n",
    "!gsutil -m cp -r $WORKSPACE_BUCKET/classes_2026-01-08/* classes_2026-01-14/\n",
    "# Combined points classes\n",
    "!gsutil -m cp -r $WORKSPACE_BUCKET/combined_classes_2026-01-08/* classes_2026-01-14/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls classes_2025-12-20/exposures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_parquet('classes_2025-12-15/exposures/calibrated_2025-12-08_BRCA2_unpublished.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_dir = Path(f'classes_{CLASSES_VERSION}')\n",
    "\n",
    "or_estimate_dicts = {}\n",
    "\n",
    "for exposures_path in tqdm((common_dir / 'exposures').iterdir()):\n",
    "    \n",
    "    try:\n",
    "        af_map_path = common_dir / 'af_maps' / exposures_path.name\n",
    "        clinvar_map_path = common_dir / 'clinvar_maps' / exposures_path.name\n",
    "\n",
    "        #display(exposures_path)\n",
    "\n",
    "        clinvar_big_df = pd.read_parquet(clinvar_map_path)\n",
    "        af_map = pd.read_parquet(af_map_path)\n",
    "\n",
    "        #display(clinvar_big_df)\n",
    "\n",
    "        #display(af_map)\n",
    "\n",
    "        #display(pd.read_parquet(exposures_path))\n",
    "\n",
    "        for (dataset, gene, classifier, classification), exposure_df in tqdm(\n",
    "            pd.read_parquet(exposures_path).groupby(\n",
    "                ['Dataset', 'Gene', 'Classifier', 'Classification'],\n",
    "                group_keys=True\n",
    "            )\n",
    "        ):\n",
    "            \n",
    "            # Patch in splice measure info\n",
    "            exposure_df['Measures splicing'] = (splicing_map.get(dataset) == 'Yes')\n",
    "            \n",
    "            #display(exposure_df)\n",
    "            if gene not in cohorts.gene_cohort_map:\n",
    "                print(f'{gene} has no cohort definition')\n",
    "                continue\n",
    "            \n",
    "            # Grab cohort\n",
    "            gene_cohort = cohorts.gene_cohort(gene)\n",
    "            \n",
    "            # Init aux info\n",
    "            aux_info_dict = {\n",
    "                'Case inclusion phenotypes': list(data.gene_phenotypes[gene][0]),\n",
    "                'Control exclusion phenotypes': list(data.gene_phenotypes[gene][1]),\n",
    "                'ENSG': ensg_map[gene]\n",
    "            }\n",
    "\n",
    "            # Extract aux info\n",
    "            for aux_info in ['Measures splicing', 'Data Version']:\n",
    "                if aux_info in exposure_df.columns:\n",
    "                    info_val = exposure_df[aux_info].iloc[0]\n",
    "                    if not (exposure_df[aux_info] == info_val).all():\n",
    "                        print(f'Warning: multiple values for {aux_info} encountered: {exposure_df}')\n",
    "                    aux_info_dict[aux_info] = info_val\n",
    "\n",
    "            # Filter splicing if needed\n",
    "            if not aux_info_dict.get('Measures splicing'):\n",
    "                splice_carriers = pd.read_parquet(f'{BUCKET}/spliceAI_2025-12-10/splice_carriers_{gene}.parquet')\n",
    "                if not splice_carriers.empty:\n",
    "                    gene_cohort = gene_cohort[\n",
    "                        ~gene_cohort.index.isin(pd.to_numeric(splice_carriers.s, downcast='integer'))\n",
    "                    ]\n",
    "                \n",
    "                # Hard code splice filter threshold for now since we know it\n",
    "                aux_info_dict['SpliceAI filter max'] = 0.2\n",
    "                        \n",
    "            # The real work\n",
    "            clinvar_class_map = clinvar_big_df[\n",
    "                (clinvar_big_df['Dataset'] == dataset) &\n",
    "                (clinvar_big_df['Gene'] == gene) &\n",
    "                (clinvar_big_df['Classifier'] == classifier) &\n",
    "                (clinvar_big_df['Classification'] == classification)\n",
    "            ].set_index('Variant')['Clinvar significance']\n",
    "\n",
    "            or_estimate_dicts[\n",
    "                (dataset, gene, classifier, classification)\n",
    "            ] = odds.estimate_logOR(\n",
    "                pd.to_numeric(exposure_df.person_id, downcast='integer'),\n",
    "                cohorts.gene_cohort(gene)\n",
    "            ) | odds.collect_variant_stats(\n",
    "                pd.to_numeric(exposure_df.person_id, downcast='integer'),\n",
    "                cohorts.gene_cohort(gene),\n",
    "                exposure_df.variants,\n",
    "                af_map['AF'],\n",
    "                clinvar_class_map\n",
    "            ) | aux_info_dict\n",
    "        \n",
    "    except:\n",
    "        print(f\"Error on {exposures_path}\")\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df = pd.DataFrame.from_dict(or_estimate_dicts, orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "or_estimates_df.to_parquet(f'{BUCKET}/or-estimates/raw-or-estimates-{CLASSES_VERSION}.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df['Few samples'] = or_estimates_df['Cases with variants'] < 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Round population count up to a multiple of 20\n",
    "or_estimates_df[\n",
    "    ['Cases with variants', 'Controls with variants', 'Cases without variants', 'Controls without variants']\n",
    "] = -(\n",
    "    -or_estimates_df[\n",
    "        ['Cases with variants', 'Controls with variants', 'Cases without variants', 'Controls without variants']\n",
    "    ] // 20\n",
    ") * 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill missing ClinVar classes as zero-counts:\n",
    "or_estimates_df[\n",
    "    or_estimates_df.columns[or_estimates_df.columns.str.contains('ClinVar')]\n",
    "] = or_estimates_df[\n",
    "    or_estimates_df.columns[or_estimates_df.columns.str.contains('ClinVar')]\n",
    "].fillna(0).apply(lambda x: pd.to_numeric(x, downcast='integer'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set index names\n",
    "or_estimates_df.index.set_names(['Dataset', 'Gene', 'Classifier', 'Classification'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "or_estimates_df.to_parquet(f'or-estimates-{CLASSES_VERSION}.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "or_estimates_df.to_parquet(f'{BUCKET}/or-estimates/or-estimates-{CLASSES_VERSION}.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "raise RuntimeError(\"Review below cells before running\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df = pd.read_parquet(f'or-estimates-{CLASSES_VERSION}.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df['ENSG'] = or_estimates_df.index.to_frame()['Gene'].map(ensg_map)\n",
    "or_estimates_df['Case inclusion phenotypes'] = or_estimates_df.index.to_frame()['Gene'].map(\n",
    "    {\n",
    "        gene: list(data.gene_phenotypes[gene][0])\n",
    "        for gene in data.gene_phenotypes\n",
    "    }\n",
    ")\n",
    "or_estimates_df['Control exclusion phenotypes'] = or_estimates_df.index.to_frame()['Gene'].map(\n",
    "    {\n",
    "        gene: list(data.gene_phenotypes[gene][1])\n",
    "        for gene in data.gene_phenotypes\n",
    "    }\n",
    ")\n",
    "or_estimates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make classification styling consistent\n",
    "# Do this by replacing underscores with spaces and capitalizing\n",
    "# Explicitly manage labels containing acronyms\n",
    "explicit_map = {\n",
    "    'possiblyLOF': 'Possibly LOF',\n",
    "    'possiblyWT': 'Possibly WT',\n",
    "    'moderate LOF': 'Moderate LOF',\n",
    "    'severe LOF': 'Severe LOF'\n",
    "}\n",
    "\n",
    "or_estimates_df['Classification'] = or_estimates_df['Classification'].where(\n",
    "    or_estimates_df['Classification'].isin({\n",
    "        'FUNC', 'INT', 'LOF', 'AF < 0.001', 'GOF', 'WT',\n",
    "        'possiblyLOF', 'possiblyWT', 'moderate LOF', 'severe LOF'\n",
    "    }),\n",
    "    or_estimates_df['Classification'].str.replace('_', ' ').str.capitalize()\n",
    ").mask(\n",
    "    or_estimates_df['Classification'].isin(explicit_map),\n",
    "    or_estimates_df['Classification'].map(explicit_map)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save estimates\n",
    "\n",
    "or_estimates_df.to_csv(f'{BUCKET}/WIP/or-estimates_2025-06-09.csv.gz', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Local save\n",
    "\n",
    "or_estimates_df.to_csv(f'or-estimates_2025-06-09.csv.gz', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "or_estimates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
